### ISSUE TITLE
Anne Pro 2 support?

### ISSUE THREAD

**Issue by jedahan** [ah-/anne-key#70]
I know it was just announced, but wondering if we could have a tracking issue for what chipset it is, etc.

**Comment by hdhoang** [comment]
sure thing! Such info would fit nicely in the book, even  if we don't have any code for it.  I'll be on the look out for them

**Comment by jedahan** [comment]
Okay I ordered one, should be here in a few days, will see what the differences are

**Comment by therealprof** [comment]
I also ordered one last Saturday from China. Although it was on backorder very much to my surprise it already arrived today in Germany...

**Comment by jedahan** [comment]
Got mine in too, what commands would be helpful? lspci? lsusb? opening it up and taking pictures?

**Comment by ah-** [comment]
Exciting! The latter, close up pictures of the chips so we can see if they're the same.

**Comment by MvEerd** [comment]
The included images were posted in the Discord by Flashquark, the chip seems to be a HT32F52342
Higher quality images of the PCB would be much appreciated though!

[HT32F52342-52_Datasheetv120.pdf](https://github.com/ah-/anne-key/files/2341225/HT32F52342-52_Datasheetv120.pdf)

<details>
  <summary>Click to see images</summary>
  ![anne_pro_2_left](https://user-images.githubusercontent.com/12560315/44931022-22872200-ad15-11e8-8e7f-fe21e53e7a45.jpg)
![anne_pro_2_right](https://user-images.githubusercontent.com/12560315/44931023-22872200-ad15-11e8-93b6-27c6bb1b7b64.JPG)

</details>

**Comment by jedahan** [comment]
![img_20180831_161207](https://user-images.githubusercontent.com/32407/44933947-10b77600-ad39-11e8-90e7-666cd14adb1c.jpg)
![img_20180831_161227](https://user-images.githubusercontent.com/32407/44933948-10b77600-ad39-11e8-953a-42d4cd645297.jpg)
![img_20180831_161300](https://user-images.githubusercontent.com/32407/44933949-10b77600-ad39-11e8-8f44-8865c33f5d8c.jpg)
![img_20180831_161324](https://user-images.githubusercontent.com/32407/44933950-11500c80-ad39-11e8-9007-c4665cae87a3.jpg)
![img_20180831_161328](https://user-images.githubusercontent.com/32407/44933951-11500c80-ad39-11e8-9852-fb5b351958fe.jpg)

**Comment by jedahan** [comment]
hmm these aren't as good as the ones you already have

**Comment by jedahan** [comment]
It does look like there is something like DFU mode if you hold esc while plugging in the usb c. Not sure if we can inspect that, maybe with a facedancer or something.

**Comment by ah-** [comment]
Got a pic of the bluetooth chip? It's under the little metal shield on the left of the board.

**Comment by therealprof** [comment]
Hm, not sure I want to remove the can... There's so much plastic on the board and I'm not not I can protect that...

**Comment by therealprof** [comment]
Lol, the can is only hold in place by tabs, it's a CC2541. Need a photo? I'd have to fire up the microscope then, it's barely imageable using a camera.

**Comment by ah-** [comment]
Cool, that's all I wanted to know :)

CC2541 is the same chip as on the previous Anne.

**Comment by therealprof** [comment]
Here you go: It's as high-res as I can do right now and best quality GH would allow me to upload. ;)

![left](https://user-images.githubusercontent.com/3321888/44935134-cf8e8880-ad6f-11e8-80d4-18e8a46207d6.jpg)
![middle](https://user-images.githubusercontent.com/3321888/44935152-dc12e100-ad6f-11e8-8702-6e5f256fa7af.jpg)
![right](https://user-images.githubusercontent.com/3321888/44935161-e46b1c00-ad6f-11e8-8f0a-674036cf90ce.jpg)

**Comment by therealprof** [comment]
@ah- Is the BT chip really that relevant for anne-key?

**Comment by therealprof** [comment]
Here's the rest of the chips close up...

![U2](https://user-images.githubusercontent.com/3321888/44935644-e1712b00-ad71-11e8-8332-422a3b5c2a23.JPG)
![U4](https://user-images.githubusercontent.com/3321888/44935645-e209c180-ad71-11e8-9be9-33ca9086795a.JPG)
![U6](https://user-images.githubusercontent.com/3321888/44935646-e209c180-ad71-11e8-863a-4c0d80e0cccd.JPG)
![U8](https://user-images.githubusercontent.com/3321888/44935647-e209c180-ad71-11e8-8d32-bd15eea07b2b.JPG)

**Comment by therealprof** [comment]
My excitement of the Anne Pro 2 is a bit dampened. Not sure whether the Anne Pro is the same, but the 2 is really quirky in BT operation and the key mapping is rubbish (the problem being they should either implement mapping keycodes for other layouts than US or preferably add a layout indication the HID profile as defined in HID Review Request 42) which makes it pretty much unusable for for me at the moment.

**Comment by SirWindfield** [comment]
@therealprof I was thinking about getting the Anne Pro / Anne Pro 2 to tinker around with the firmware a bit. What do you mean by "the keymapping is garbish"?

**Comment by therealprof** [comment]
@SirWindfield They don't indicate the keyboard layout in the communication so it will automatically use the keyboard layout which is very annoying if you're on a laptop with a foreign layout. Should be trivial to indicate a US layout in the HID profile to allow seamless swapping between keyboards...

I was kind of shocked to learn to keyboards still use those idiotic arbitrarily chosen fixed scancodes instead of just sending the  translated characters...

**Comment by ah-** [comment]
@SirWindfield, if you want to tinker around you might want to get a Anne Pro 1, STM32 has better support for open source. Also Cortex M3 instead of M0+ :)

**Comment by ah-** [comment]
Oh yeah, there are tons. However, they usually don't have Bluetooth and tend to be a bit more expensive. If you just want to play around a bit I'd recommend getting one with QMK, if you like Rust or love reverse engineering get the Anne Pro.

**Comment by SirWindfield** [comment]
> Unless you want to help with bringup, then get the 2.

@ah- By bringup you mean helping out writing the firmware right? Not sure what is means :/
The thing is, I can't seem to find any Anne V1 anymore. They are all sold out sadly.

**Comment by ah-** [comment]
Basically figuring out what tools to use to flash / debug the chip, how to configure them, reading the datasheet to figure out how to toggle/read pins etc. So the rather low level things vs. adding higher level features like keyboard shortcuts or similar.

**Comment by ah-** [comment]
Actually, I've got a spare AP 1 that I don't use, if you want it I'll send it to you for free. It's got the red switches, don't really like them and have enough keyboards.

Send me an email (see profile) with your address if interested, no strings attached.

**Comment by hdhoang** [comment]
> Basically figuring out what tools to use to flash / debug the chip, how to configure them, reading the datasheet to figure out how to toggle/read pins etc.

I see you have a todos.txt in gitignore, could you track it here too? We can then tease it out into the book sections.

**Comment by ah-** [comment]
I guess this thread is enough tracking for now, getting it to work won't be super easy since I haven't immediately found great resources for the Holtek Cortex M0 chips.

https://github.com/ChaoticConundrum/openocd-ht32 seems to add support to openocd, but it's in a fork not in mainline etc. The nice thing about the stm32 in the original Anne Pro was that it has great support in the open source ecosystem.

**Comment by felipesere** [comment]
Would it be possible to use https://github.com/rust-embedded/cortex-m-quickstart
and the details from https://www.holtek.com/documents/10179/116711/HT32F52342_52_Datasheetv150.pdf to get started with a compilation?
I don't have an Anne2 myself, but I am curious if I can use an Anne2 and this repo to create a couple of layouts that allow me to input US keystrokes on a british OS :)

**Comment by MatiasPuhakka** [comment]
What's the latest status on Anne Pro 2 support?

**Comment by si973mo** [comment]
Can you help me with a problem with my Anne pro2 the backlight will not work when I press fn2 and 9 it will not show up any colours

**Comment by MycroftHolmes1989** [comment]
Has anyone tried dabbling with this anymore? Ducky One 2 Mini now has a QMK port, and I feel like such a fool for choosing Anne Pro 2 over it.

**Comment by DonaldTsang** [comment]
It's 2020 and I am basically hoping for the same thing for cheap 61 key keyboards

**Comment by hargoniX** [comment]
So I've been looking into implementing a HAL for the HTF32F523 series of chips in rust (I guess this would be the first step for this goal?). And got myself some hardware (specifically HT-ESK32-30501). Now I'm currently struggling to find any open source software that is able to communicate with the on board debugger, does someone here have any suggestion for that or do I have to implement support for it in something popular like openocd?

**Comment by therealprof** [comment]
Instead of going for OpenOCD I'd probably go for https://probe.rs/ first.

**Comment by hargoniX** [comment]
Oh wow I didnt know it had gotten that far since i last looked at it! Definitely will do!

**Comment by hargoniX** [comment]
@therealprof https://github.com/probe-rs/probe-rs/issues/172 just as a tracker for progress on this.

**Comment by theobarberbany** [comment]
@hargoniX I'd be happy to donate some time if you get anywhere with that! I've not got any real experience, but would be willing to try learn :)

I've been toying with the idea of trying but also don't really know where to start...

**Comment by hargoniX** [comment]
Quiiiiiick update, I've just managed to get cargo-flash to flash some random ELF onto my dev board so I'll be landing a PR for that in cargo-flash soon.

@theobarberbany First of we'd have to look into a PAC, https://github.com/rmsthebest/ht32f523x2 is a thing but I am not sure how up to date/maintained etc it is. Usually though the process of building a PAC is mostly automated as one can run svd2rust over a manufacturer supplied SVD file and generate lots of (hopefully) correct code from that. Afterwards we'd be looking into writing an actual HAL + presumably USB support (not defined in embedded-hal afaik). And then on top of that a full keyboard firmware like this one.

**Comment by philipmnelson** [comment]
@hargoniX Any updates? Is there anything I could possibly do to help?

**Comment by hargoniX** [comment]
So I created https://github.com/ht32-rs/ and yesterday published my own set of PACs (for all Holtek chips), see [here](https://crates.io/search?page=1&per_page=10&q=ht32) and [here](https://crates.io/crates/ht32f5xxxx). What we'd have to do now is look into building a HAL for the ht32f52342_52 variant of the second crate. Which essentially means reading [this](https://www.holtek.com/documents/10179/187541/HT32F52342-52_UserManualv130.pdf) document and writing code that implements embedded-hal traits for it.

If we want to use this HAL for Anne Pro 2 firmware ASAP it would be very useful to know which peripherals we should focus on so somebody should make a list of that and file an issue for that over at https://github.com/ht32-rs/ht32f5xxxx-hal

I was gonna look into writing code for the clock setup next weekend (and probably beyond that, dunno how long that takes me) and after that look into implementing the necessary peripherals so it would certainly also be useful if someone wanted to help with the peripherals (especially with USB of course).

Apart from that nothing comes to mind @philipmnelson

**Comment by hargoniX** [comment]
And for another update, I'm done with the clock implementations as well as the GPIO implementation (see https://github.com/ht32-rs/ht32f5xxxx-hal) and will move on to the peripheral implementations for the chips posted far above by now soon :tm:

**Comment by ah-** [comment]
Wow! Does flashing and debugging work?

**Comment by hargoniX** [comment]
On the HT32 chip it has been working since https://github.com/ah-/anne-key/issues/70#issuecomment-596231100 was closed^^ Now it can already blink LEDs \o/

**Comment by hargoniX** [comment]
And for another update, I just got SPI working, yesterday got I2C working-ish so the only two things that's left now is UART and USB, how hard can USB be 😉

**Comment by MycroftHolmes1989** [comment]
Quick question: once the USB part is implemented, will it be possible to run QMK on it?

**Comment by hargoniX** [comment]
QMK is not at all related to what I am doing here, I'm implementing the Rust abstractions required to do sensible embedded development on the Holtek chip, QMK is a C firmware project that has absolutey 0 idea on how to deal with the Rust abstractions what so ever.

TLDR; QMK is a different language and compatability between Rust embedded and it = 0 -> No

**Comment by MycroftHolmes1989** [comment]
As far as I remember, there was someone who ported a few functionalities of QMK to Rust. I will try to find that project. If anything remotely close works with Anne pro 2, it will be a blessing.

**Comment by hargoniX** [comment]
U(S)ART now works as well, will look into USB in the next days/weeks and once I got that done we should be more or less ready to go with porting!

**Comment by Codetector1374** [comment]
For a QMK port you will need to port ChibiOS. I have a port of ChibiOS for HT32F165x, at github.com/OpenDucky
I recently ported QMK to a Ducky keyboard which uses the HT32F1654. It looks very close to the F5 series that they use on the Anne Pro.

**Comment by MycroftHolmes1989** [comment]
> For a QMK port you will need to port ChibiOS. I have a port of ChibiOS for HT32F165x, at [github.com/OpenDucky](https://github.com/OpenDucky)

The question is, does Anne Pro 2 have enough RAM and ROM for a working version of ChibiOS + QMK? Otherwise, it might be worthwhile to cherry-pick the features we need and port them to Rust.

**Comment by Codetector1374** [comment]
yes. QMK + Chibi uses about 20K of rom which Anne pro 2 has 64 and ~6K of ram in my case but that’s having an 16bit 16*8 led matrix which you don’t need. So you can probably do it in ~5K which Anne pro 2 has 8 I’m pretty sure.

**Comment by Codetector1374** [comment]
I am currently reversing the firmware that is on stock Anne pro 2. Hoping to be able to dump the original boot loader though a custom “firmware”.

The goal is to be able to keep the old boot loader with the new firmware. Since it only uses the lower 16K and keeps the ability to roll back to stock if needed. I think that’d be nicer than having to using SWD tool to wipe the chip.

**Comment by Codetector1374** [comment]
Honestly I am thinking about making a rust port of QMK. They really don't use ChibiOS for real. They use it mainly as a good HAL. afaik they run all of their code in the ChibiOS "idle" task. Which is kind of under utilization of an RTOS. But the upside is that any user can easily expand functionality by creating a new thread and not have to worry about the existing core QMK code.

**Comment by ah-** [comment]
> I am currently reversing the firmware that is on stock Anne pro 2. Hoping to be able to dump the original boot loader though a custom “firmware”. The goal is to be able to keep the old boot loader with the new firmware. Since it only uses the lower 16K and keeps the ability to roll back to stock if needed. I think that’d be nicer than having to using SWD tool to wipe the chip.

In the old anne you could use the default bootloader to flash your own firmware by just locating it at the right offset hopefully that works on the new one as well.

**Comment by Codetector1374** [comment]
> > I am currently reversing the firmware that is on stock Anne pro 2. Hoping to be able to dump the original boot loader though a custom “firmware”. The goal is to be able to keep the old boot loader with the new firmware. Since it only uses the lower 16K and keeps the ability to roll back to stock if needed. I think that’d be nicer than having to using SWD tool to wipe the chip.
> 
> In the old anne you could use the default bootloader to flash your own firmware by just locating it at the right offset hopefully that works on the new one as well.

Well the problem is here that they don't use standard USB-DFU. So we need to figure out what protocol they use. The firmware is loaded at 0x4000. Hence the bootloader uses the lower 16K of flash.

**Comment by Codetector1374** [comment]
Obins Kit is very stupid in the sense that it uses a HTTP server to download the firmware. So my plan right now is to write a simple USB DFU thing like this [HT32 USBDFU](https://github.com/OpenDucky/ht32-usb-dfu) base it at `0x4000` and use local DNS to trick it into downloading the "DFU" as the firmware. This will allow us to dump the protected flash (mainly the bootloder) so we can figure what they do.

**Comment by Codetector1374** [comment]
Okay I started a Github Org tracking the re-firmware progress. https://github.com/OpenAnnePro
I have the original bootloader dump uploaded there.

**Comment by Codetector1374** [comment]
More progress! 
[AnnePro2-Tools](https://github.com/OpenAnnePro/AnnePro2-Tools)
This is a small tool I wrote that allow you to flash custom firmware into the main MCU(U2). In theory you can change it to flash the other two by simply changing the target param. But that work is to be tested. Enjoy :smile:

**Comment by MycroftHolmes1989** [comment]
> U(S)ART now works as well, will look into USB in the next days/weeks and once I got that done we should be more or less ready to go with porting!

I hope we get a Rust implementation of the firmware too, for those who don't want to use QMK.

**Comment by hargoniX** [comment]
Im still looking into USB, never implemented a USB peripheral in Rust though (and the way Holtek's USB peripheral was built also makes it a bit hard to generalize into the traits from usb-device)
